### PR TITLE
[SYCLomatic] Support migration of cub::DeviceReduce::Sum, cub::DeviceScan::InclusiveSum, cub::DeviceScan::ExclusiveSum and cub::DeviceSelect::Flagged

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/APINamesCUB.inc [hunk 1]
```diff
@@ -0,0 +1,60 @@
+//===--------------- APINamesComplex.inc ----------------------------------===//
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/APINamesCUB.inc]: APINamesCUB.inc

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]
LGTM

File: clang/lib/DPCT/CubRule.cpp [hunk 2]
```diff
@@ -0,0 +1,1354 @@
+//===--------------- CubRule.cpp ------------------------------------------===//
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CubRule.cpp]: how about CUBAPIMigration.cpp,  keep the style alined with other file already exist, like: BLASAPIMigration.cpp, FFTAPIMigration.cpp

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 3]
```diff
@@ -2118,6 +2119,24 @@ createInsertHeaderFactory(
   return createInsertHeaderFactory(Header, std::move(Input));
 }
 
+std::pair<std::string, std::shared_ptr<CallExprRewriterFactoryBase>>
+createRemoveCubTempStorageFactory(
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CallExprRewriter.cpp:2123]: if it is cub specific,  then how about put them into a seperate cub util file?

File: clang/lib/DPCT/CubRule.cpp [hunk 4]
```diff
@@ -0,0 +1,1354 @@
...
+  } else if (isCubDeviceFuncName(FuncName)) {
+    processDeviceLevelFuncCall(CE, FuncCallUsed);
+  }
+}
+
+void CubRule::processBlockLevelMemberCall(const CXXMemberCallExpr *BlockMC) {
+  if (!BlockMC || !BlockMC->getMethodDecl()) {
+    return;
+  }
+  std::string Repl, NewFuncName, ParamList, InitRepl, OpRepl, Indent,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CubRule.cpp]: For functinality of this file,  I see some function is moved from [clang/lib/DPCT/ASTTraversal.cpp], so please ask old author "wang, hao3" to have a review.

File: clang/lib/DPCT/cmake/libDPCT.cmake [hunk 5]
```diff
@@ -32,6 +32,7 @@ macro(build_lib_dpct)
     TypeLocRewriters.cpp
     QueryApiMapping.cpp
     LIBCUAPIMigration.cpp
+    CubRule.cpp
```
Comment (zhimingwang36) [inline clang/lib/DPCT/cmake/libDPCT.cmake]: filename need to be updated to align with other filenames.

File: clang/test/dpct/cub/devicelevel/device_scan_flagged.cu [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  int host_in[] =       {1, 2, 3, 4, 5};
+  int host_flagged[] =  {0, 1, 0, 1, 0};
+  int host_out[5];
+  cudaMalloc((void **)&device_in, n * sizeof(int));
+  cudaMalloc((void **)&device_out, n * sizeof(int));
+  cudaMalloc((void **)&device_flagged, n * sizeof(int));
+  cudaMalloc((void **)&device_select_num, sizeof(int));
+  cudaMemcpy(device_in, (void *)host_in, sizeof(host_in), cudaMemcpyHostToDevice);
+  cudaMemcpy(device_flagged, host_flagged, sizeof(host_flagged), cudaMemcpyHostToDevice);
+  cub::DeviceSelect::Flagged(device_tmp, n_device_tmp, device_in, device_flagged, device_out, device_select_num, n), 0;
```
Comment (zhimingwang36) [inline clang/test/dpct/cub/devicelevel/device_scan_flagged.cu]: how about update the test case filename to device_select_flagged.cu?

**Review by zhimingwang36 — CHANGES_REQUESTED** [review]
in commit msg, it should be cub::DeviceScan::ExclusiveSum instead of  cub::Device::ExclusiveSum, right?

File: clang/lib/DPCT/CubRule.cpp (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/CubRule.cpp]: +1

File: clang/lib/DPCT/cmake/libDPCT.cmake (hunk 5 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/cmake/libDPCT.cmake]: +1

File: clang/test/dpct/cub/devicelevel/device_scan_flagged.cu (hunk 6 above)
Comment (yihwang-nv) [reply clang/test/dpct/cub/devicelevel/device_scan_flagged.cu]: +1

File: clang/lib/DPCT/CallExprRewriter.cpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/CallExprRewriter.cpp:2123]: This function just a wrapper and it only used for macro in CallExprRewriter.cpp, the main logic is implemented in CUBAPIMigration.cpp

File: clang/lib/DPCT/CubRule.cpp (hunk 4 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/CubRule.cpp]: +1

**Comment by yihwang-nv** [comment]
> in commit msg, it should be cub::DeviceScan::ExclusiveSum instead of cub::Device::ExclusiveSum, right?

Yes, thanks!

**Review by intwanghao — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**